drivers/adc/ad796x/ad796x.c

```diff
@@ -0,0 +1,297 @@
+
+	return 0;
+}
+
+/**
+ * @brief Initialize the device gpios.
+ * @param device - The device structure.
+ * @param init_param - The structure that contains the device initial
+ * 		       parameters.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int ad796x_gpio_init(struct ad796x_dev *dev,
+			    struct ad796x_init_param *init_param)
+{
+	int ret, i;
+
+	struct no_os_gpio_init_param *params[] = {
+		init_param->gpio_adc_en3_fmc_ip,
+		init_param->gpio_adc_en2_fmc_ip,
+		init_param->gpio_adc_en1_fmc_ip,
+		init_param->gpio_adc_en0_fmc_ip,
+	};
+
+	struct no_os_gpio_desc *desc[] = {
+		dev->gpio_adc_en3_fmc,
+		dev->gpio_adc_en2_fmc,
+		dev->gpio_adc_en1_fmc,
+		dev->gpio_adc_en0_fmc,
+	};
+
+	uint8_t value[] = {
+		ad796x_modes[init_param->mode].en3,
+		ad796x_modes[init_param->mode].en2,
+		ad796x_modes[init_param->mode].en1,
+		ad796x_modes[init_param->mode].en0,
+	};
+
+	for (i = 0; i < AD796X_NUM_GPIOS; i++) {
+		ret = no_os_gpio_get_optional(&desc[i], params[i]);
+		if (ret)
+			goto error;
+
+		if (!desc[i])
+			continue;
+
+		ret = no_os_gpio_direction_output(desc[i], value[i]);
+		if (ret)
+			goto error;
+	}
+
+	return 0;
+error:
+	ad796x_gpio_remove(dev);
+	return ret;
+}
+
+/**
+ * @brief Initialize the device.
+ * @param device - The device structure.
+ * @param init_param - The structure that contains the device initial
+ * 		       parameters.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad796x_init(struct ad796x_dev **device,
+		struct ad796x_init_param *init_param)
+{
+	struct ad796x_dev *dev;
+	int ret;
+
+	dev = (struct ad796x_dev *)no_os_calloc(1, sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
+
+	ret = ad796x_gpio_init(dev, init_param);
+	if (ret)
+		goto error;
+
+	ret = axi_clkgen_init(&dev->clkgen, init_param->clkgen_ip);
+	if (ret) {
+		pr_err("error: axi_clkgen_init error: %d\n", ret);
+		goto error;
+	}
+
+	ret = axi_clkgen_set_rate(dev->clkgen, 125000000);
+	if (ret) {
+		pr_err("error: axi_clkgen_set_rate error: %d\n", ret);
+		goto error;
+	}
+
+	ret = no_os_pwm_init(&dev->axi_pwm_0, init_param->axi_pwm_0_ip);
+	if (ret) {
+		pr_err("error: no_os_pwm_init error: %d\n", ret);
+		goto error;
+	}
+
+	ret = no_os_pwm_init(&dev->axi_pwm_1, init_param->axi_pwm_1_ip);
+	if (ret) {
+		pr_err("error: no_os_pwm_init error: %d\n", ret);
+		goto error;
+	}
+
+	ret = axi_adc_init(&dev->ad796x_core, init_param->ad796x_core_ip);
+	if (ret) {
+		pr_err("axi_adc_init error: %d\n", ret);
+		goto error;
+	}
+
+	ret = axi_dmac_init(&dev->axi_dmac, init_param->dmac_ip);
+	if (ret) {
+		pr_err("axi_dmac_init error: %d\n", ret);
+		goto error;
+	}
+
+	*device = dev;
+
+	return 0;
+
+error:
+	ad796x_remove(dev);
+	return ret;
+}
+
+/**
+ * @brief Remove the device and release resources.
+ * @param dev - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad796x_remove(struct ad796x_dev *dev)
+{
+	int ret;
+
+	if (dev->ad796x_core) {
```

Comment:
the only thing I would check for null here is dev.
freeing a null pointer is allowed